sdk/servicebus/service-bus/src/receivers/sessionReceiver.ts

```diff
@@ -113,56 +122,68 @@ export class SessionReceiverImpl<ReceivedMessageT extends ReceivedMessage | Rece
    * @throws Error if an open receiver is already existing for given sessionId.
    */
   private constructor(
+    messageSession: MessageSession,
     context: ClientEntityContext,
     public receiveMode: "peekLock" | "receiveAndDelete",
     private _sessionOptions: CreateSessionReceiverOptions,
     retryOptions: RetryOptions = {}
   ) {
     throwErrorIfConnectionClosed(context.namespace);
     this._context = context;
+    this._sessionOptions;
     this.entityPath = this._context.entityPath;
     this._retryOptions = retryOptions;
+    this._messageSession = messageSession;
+    // By this point, we should have a valid sessionId on the messageSession
+    // If not, the receiver cannot be used, so throw error.
+    if (messageSession.sessionId == undefined) {
+      const error = new Error("Something went wrong. Cannot lock a session.");
+      log.error(`[${context.namespace.connectionId}] %O`, error);
+      throw error;
+    }
+    this.sessionId = messageSession.sessionId;
+    delete this._context.expiredMessageSessions[messageSession.sessionId];
+  }
 
-    if (this._sessionOptions.sessionId) {
-      this._sessionOptions.sessionId = String(this._sessionOptions.sessionId);
+  static async createInitializedSessionReceiver<
+    ReceivedMessageT extends ReceivedMessage | ReceivedMessageWithLock
+  >(
+    context: ClientEntityContext,
+    receiveMode: "peekLock" | "receiveAndDelete",
+    sessionOptions: CreateSessionReceiverOptions,
+    retryOptions: RetryOptions = {}
+  ): Promise<SessionReceiver<ReceivedMessageT>> {
+    context.isSessionEnabled = true;
+    if (sessionOptions.sessionId) {
+      sessionOptions.sessionId = String(sessionOptions.sessionId);
 
       // Check if receiver for given session already exists
       if (
-        this._context.messageSessions[this._sessionOptions.sessionId] &&
-        this._context.messageSessions[this._sessionOptions.sessionId].isOpen()
+        context.messageSessions[sessionOptions.sessionId] &&
+        context.messageSessions[sessionOptions.sessionId].isOpen()
       ) {
         const errorMessage = getOpenSessionReceiverErrorMsg(
-          this._context.entityPath,
-          this._sessionOptions.sessionId
+          context.entityPath,
+          sessionOptions.sessionId
```

Comment:
If we don't throw this error and continue to create the message session, will the service throw `SessionCannotBeLocked` exception? If yes, then I wonder if we should just let that happen instead us throwing a custom error here

Reply:
Okay, I found some discrepancies here by testing with creating multiple session receivers with same session-id.
- This if-block is never hit and the `await MessageSession.create()` would throw from `MessageSession._init()`
    > MessagingError: Received sessionId 'undefined' does not match the provided sessionId 'my-session' while creating the receiver 'partitioned-queue-sessions-4c65bb4c-bb7b-9043-b35b-8e9277d5745b'.
- Reason being the clientEntityContext is different for

Reply:
> Reason being the clientEntityContext is different when the createSessionReceiver is called for the second time.

I believe the reason behind this must be that, in track-1, each entity had its own client and the clientEntityContext was unique for the entity. 
Now, that we have moved away from that model, there could be many `clientEntityContext`s for an entity. 

Without touching the `ClientEntityContext` code, I can get the if-block working by modifying the code to look for all the `clientEntityContext`s that have the same `entityPath` and then look for message sessions.

Reply:
With this https://github.com/Azure/azure-sdk-for-js/pull/8849/commits/66a25244db271275a46793fa45d4692efd7ed4ce, the if-check works as expected.

### Open Questions
- Keep unique clientEntityContext for an entity?
- Should the if-block that checks for existing messageSessions with the current session-id be removed?
   - If the above if-block is removed, the service doesn't throw. We throw `Received sessionId 'undefined' does not match the provided sessionId` from `MessageSession._init()`. This may be fine too though I like checking the existing messageSessions first.
- Update the error thrown `Received sessionId 'undefined' does not match the provided sessionId` from `MessageSession._init()` to `SessionCannotBeLocked` error?

Reply:
I really don't like each sender and receiver having its own client entity context. It is an overkill. I just put out #9710 to make it so that this context has a 1:1 mapping with the entity i.e queue/topic/subscription. That should simplify things here.

I am really surprised that the service does not throw a `SessionCannotBeLocked` error if we try to create a session for a sessionId that already exists. We can follow up on this in a separate PR

Reply:
With https://github.com/Azure/azure-sdk-for-js/pull/9766, we no longer need to deal with the checks on the clientEntityContext being discussed in this thread.